### PR TITLE
feat[warg]: Operator logs

### PR DESCRIPTION
**Pull request by Kylebrown9**
* Added pbjson_build to enable proto3 JSON mapping.
* Normalized case for algorithms (e.g. SHA256) to lowercase (e.g. sha256).
* Implemented operator log protobuf, parsing, and validation.
* Centralized operator and package protos.
* Added JSON-file-based test suite system for log validation.

### CODE REVIEW

File: crates/warg/src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,154 @@
...
+        Ok(Envelope {
+            contents,
+            content_bytes,
+            key_id,
+            signature,
+        })
+    }
+}
+
+type ErrorBox = Box<dyn Error + Send + Sync + 'static>;
```
Comment (lann) [inline crates/warg/src/lib.rs:113]: Any particular reason to avoid `anyhow` here?

**Review by lann — COMMENTED** [review]
Incremental review

File: crates/warg/src/operator/mod.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+impl TryFrom<protobuf::OperatorRecord> for model::OperatorRecord {
+    type Error = ErrorBox;
+
+    fn try_from(record: protobuf::OperatorRecord) -> Result<Self, Self::Error> {
+        let prev: Option<hash::Digest> = match record.prev {
+            Some(hash_string) => Some(hash_string.parse()?),
+            None => None,
+        };
+        let version = record.version;
+        let pbjson_timestamp = record.time.ok_or_else(|| Box::new(InvalidTimestampError))?;
```
Comment (lann) [inline crates/warg/src/operator/mod.rs:30]: fwiw `anyhow::Context` lets you do `record.time.context(InvalidTimestampError)?`

File: crates/warg/src/operator/mod.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                    key_id: bob_pub.fingerprint(),
+                    permission: model::Permission::Commit,
+                },
+            ],
+        };
+
+        let first_envelope = match Envelope::signed_contents(&alice_priv, record) {
+            Ok(value) => value,
+            Err(error) => panic!("Failed to sign envelope 1: {:?}", error),
+        };
```
Comment (lann) [inline crates/warg/src/operator/mod.rs:202]: Almost exactly equivalent:

```suggestion
        let first_envelope = Envelope::signed_contents(&alice_priv, record)
                .expect("Failed to sign envelope 1");
```

File: crates/warg/src/lib.rs [hunk 4]
```diff
@@ -1,4 +1,154 @@
...
+
+/// Errors that occur in the process of parsing an envelope from bytes
+#[derive(Error, Debug)]
+pub enum ParseEnvelopeError {
+    #[error("Failed to parse the outer envelope protobuf message")]
+    ProtobufEnvelopeParseError(#[from] prost::DecodeError),
+
+    #[error("Failed to parse envelope contents from bytes")]
+    ContentsParseError(ErrorBox),
```
[the remark this answers is not in the document]
Comment (Kylebrown9) [reply crates/warg/src/lib.rs:122]: Huh, I think that used to introduce a type ambiguity. Seems to work now though.

File: crates/warg/src/lib.rs [hunk 5]
```diff
@@ -1,4 +1,154 @@
...
+
+use hash::HashParseError;
+use signature::Error as SignatureError;
+use signing::SignatureParseError;
+
+/// The envelope struct is used to keep around the original
+/// bytes that the content was serialized into in case
+/// the serialization is not canonical.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Envelope<Contents> {
```
[the remark this answers is not in the document]
Comment (Kylebrown9) [reply crates/warg/src/lib.rs:43]: Ya, this was just me being a little sloppy making things `pub` for the integration tests.

File: crates/warg/src/lib.rs (hunk 1 above)
Comment (Kylebrown9) [reply crates/warg/src/lib.rs:113]: Just unfamiliarity, `anyhow::Error` does do the trick nicely though so I've implemented that.

File: crates/warg/src/operator/mod.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+use crate::hash;
+
+pub mod model;
+pub mod validate;
+
+use crate::{protobuf, Encode, ErrorBox, Signable};
+
+// Deserialization
+
+impl TryFrom<&[u8]> for model::OperatorRecord {
```
[the remark this answers is not in the document]
Comment (Kylebrown9) [reply crates/warg/src/operator/mod.rs:14]: Added a symmetric trait to `Encode` called `Decode` for this.

File: crates/warg/src/operator/mod.rs (hunk 2 above)
Comment (Kylebrown9) [reply crates/warg/src/operator/mod.rs:31]: Sounds nice.

File: crates/warg/src/operator/mod.rs (hunk 3 above)
Comment (Kylebrown9) [reply crates/warg/src/operator/mod.rs:203]: Implemented.

**Review by lann — APPROVED** [review]

**Pull request merged**